File: scorio/eval/gpass.py
r"""Generalized pass-family evaluation metrics for binary outcomes.

Estimate stability-style pass metrics from a binary outcome matrix
:math:`R \in \{0,1\}^{M \times N}`. ``G-Pass@k`` is the all-success
threshold, ``G-Pass@k``\ :sub:`\tau` requires at least
:math:`\max(1, \lceil \tau k \rceil)` successes, and ``mG-Pass@k`` uses
the cited discrete approximation over thresholds
:math:`\tau \in [0.5, 1.0]`. Bayesian summaries compute posterior
``mu`` and ``sigma`` under a Beta model for each question's latent success
rate.

Available API
-------------------
- ``g_pass_at_k`` returns the :math:`\tau = 1` alias of ``pass_hat_k``.
- ``g_pass_at_k_tau`` returns the thresholded generalized pass metric.
- ``mg_pass_at_k`` returns the mean generalized pass metric.
- companion ``*_ci`` functions return ``(mu, sigma, lo, hi)`` using a
  normal-approximation credible interval around ``mu``.
"""

import math

import numpy as np

from ._count_score import CountScore
from ._inputs import prepare_binary_bank, validate_finite_k
from ._posterior import posterior_moments
from .pass_at_k import pass_hat_k, pass_hat_k_ci
from .utils import normal_credible_interval


def g_pass_at_k(R: np.ndarray, k: int) -> float:
    r"""
    Performance evaluation using G-Pass@k.

    Equivalent to :func:`~scorio.eval.pass_hat_k`, included for literature
    that uses the G-Pass@k naming convention for the
    :math:`\tau = 1` threshold.

    References:
        Liu, J., Liu, H., Xiao, L., et al. (2025).
        Are Your LLMs Capable of Stable Reasoning?
        *Findings of ACL 2025*, 17594--17632.
        https://doi.org/10.18653/v1/2025.findings-acl.905

    Args:
        R: :math:`M \times N` binary matrix with entries in :math:`\{0, 1\}`.
           :math:`R_{\alpha i} = 1` if trial :math:`i` for question
           :math:`\alpha` passed, 0 otherwise.
        k: Number of samples to select (:math:`1 \le k \le N`).

    Returns:
        float: The average G-Pass@k score across all :math:`M` questions.

    Notation:
        For each row :math:`\alpha`:

        .. math::

            \nu_\alpha = \sum_{i=1}^{N} R_{\alpha i} \quad \text{(number of correct samples)}

        :math:`C(a, b)` denotes the binomial coefficient :math:`\binom{a}{b}`.

    Formula:
        .. math::

            \mathrm{G\text{-}Pass@}k_\alpha = \frac{C(\nu_\alpha, k)}{C(N, k)}

        .. math::

            \mathrm{G\text{-}Pass@}k = \frac{1}{M} \sum_{\alpha=1}^{M}
                \mathrm{G\text{-}Pass@}k_\alpha

    Examples:
        >>> import numpy as np
        >>> R = np.array([[0, 1, 1, 0, 1],
        ...               [1, 1, 0, 1, 1]])
        >>> round(g_pass_at_k(R, 1), 6)
        0.7
        >>> round(g_pass_at_k(R, 2), 6)
        0.45
    """
    return pass_hat_k(R, k)


def g_pass_at_k_tau(R: np.ndarray, k: int, tau: float) -> float:
    r"""
    Performance evaluation using G-Pass@k\ :sub:`τ`.

    References:
        Liu, J., Liu, H., Xiao, L., et al. (2025).
        Are Your LLMs Capable of Stable Reasoning?
        *Findings of ACL 2025*, 17594--17632.
        https://doi.org/10.18653/v1/2025.findings-acl.905

    Args:
        R: :math:`M \times N` binary matrix with entries in :math:`\{0, 1\}`.
           :math:`R_{\alpha i} = 1` if trial :math:`i` for question
           :math:`\alpha` passed, 0 otherwise.
        k: Number of samples to select (:math:`1 \le k \le N`).
        tau: Threshold parameter :math:`\tau \in [0, 1]`. Requires at
             least :math:`\max(1, \lceil \tau \cdot k \rceil)` successes.
             When :math:`\tau = 0`, equivalent to Pass@k.
             When :math:`\tau = 1`, equivalent to Pass^k.

    Returns:
        float: The average G-Pass@k\ :sub:`τ` score across all :math:`M` questions.

    Notation:
        For each row :math:`\alpha`:

        .. math::

            \nu_\alpha = \sum_{i=1}^{N} R_{\alpha i} \quad \text{(number of correct samples)}

        :math:`C(a, b)` denotes the binomial coefficient :math:`\binom{a}{b}`.

        :math:`j_0 = \max(1, \lceil \tau \cdot k \rceil)` is the minimum
        number of successes required. The lower bound of one makes
        :math:`\tau=0` equivalent to Pass@k rather than a vacuous event.

    Formula:
        .. math::

            \mathrm{G\text{-}Pass@}k_{\tau, \alpha} = \sum_{j=j_0}^{k}
                \frac{C(\nu_\alpha, j) \cdot C(N - \nu_\alpha, k - j)}{C(N, k)}

        .. math::

            \mathrm{G\text{-}Pass@}k_\tau = \frac{1}{M} \sum_{\alpha=1}^{M}
                \mathrm{G\text{-}Pass@}k_{\tau, \alpha}

    Examples:
        >>> import numpy as np
        >>> R = np.array([[0, 1, 1, 0, 1],
        ...               [1, 1, 0, 1, 1]])
        >>> round(g_pass_at_k_tau(R, 2, 0.5), 6)
        0.95
        >>> round(g_pass_at_k_tau(R, 2, 1.0), 6)
        0.45
    """
    bank = prepare_binary_bank(R)

    if not (0.0 <= tau <= 1.0):
        raise ValueError(f"tau must be in [0, 1]; got {tau}")
    k = validate_finite_k(bank.trial_count, k)

    if tau <= 0.0:
        return CountScore.pass_at_k(k).mean(bank)

    threshold = max(1, int(math.ceil(tau * k)))
    return CountScore.threshold_at_k(k, threshold).mean(bank)


def mg_pass_at_k(R: np.ndarray, k: int) -> float:
    r"""
    Performance evaluation using mG-Pass@k.

    References:
        Liu, J., Liu, H., Xiao, L., et al. (2025).
        Are Your LLMs Capable of Stable Reasoning?
        *Findings of ACL 2025*, 17594--17632.
        https://doi.org/10.18653/v1/2025.findings-acl.905

    Args:
        R: :math:`M \times N` binary matrix with entries in :math:`\{0, 1\}`.
           :math:`R_{\alpha i} = 1` if trial :math:`i` for question
           :math:`\alpha` passed, 0 otherwise.
        k: Number of samples to select (:math:`1 \le k \le N`).

    Returns:
        float: The average mG-Pass@k score across all :math:`M` questions.

    Notation:
        For each row :math:`\alpha`:

        .. math::

            \nu_\alpha = \sum_{i=1}^{N} R_{\alpha i} \quad \text{(number of correct samples)}

        :math:`m = \lceil k/2 \rceil` is the majority threshold.
        Let :math:`X_\alpha \sim \mathrm{Hypergeometric}(N, \nu_\alpha, k)`
        be the number of correct samples among :math:`k` selections for row
        :math:`\alpha`.

    Formula:
        The cited G-Pass metric defines mG-Pass by the following discrete
        summation. It is a right-endpoint approximation to the normalized
        continuous area over :math:`\tau \in [0.5,1]`, rather than an exact
        integral identity for arbitrary (in particular, odd) :math:`k`.

        .. math::

            \mathrm{mG\text{-}Pass@}k_\alpha \approx 2 \int_{0.5}^{1.0}
                \mathrm{G\text{-}Pass@}k_{\tau, \alpha} \, d\tau

        .. math::

            \mathrm{mG\text{-}Pass@}k_\alpha = \frac{2}{k} \sum_{j=m+1}^{k}
                (j - m) \cdot P(X_\alpha = j)

        .. math::

            P(X_\alpha = j) = \frac{C(\nu_\alpha, j) \cdot C(N - \nu_\alpha, k - j)}{C(N, k)}

        .. math::

            \mathrm{mG\text{-}Pass@}k = \frac{1}{M} \sum_{\alpha=1}^{M}
                \mathrm{mG\text{-}Pass@}k_\alpha

    Examples:
        >>> import numpy as np
        >>> R = np.array([[0, 1, 1, 0, 1],
        ...               [1, 1, 0, 1, 1]])
        >>> round(mg_pass_at_k(R, 2), 6)
        0.45
        >>> round(mg_pass_at_k(R, 3), 6)
        0.166667
    """
    bank = prepare_binary_bank(R)
    k = validate_finite_k(bank.trial_count, k)
    return CountScore.mg_at_k(k).mean(bank)


def _g_pass_at_k_tau_bayes(
    R: np.ndarray, k: int, tau: float, alpha0: float = 1.0, beta0: float = 1.0
) -> tuple[float, float]:
    """Posterior mean/std for the i.i.d. G-Pass@k_τ quantity."""
    bank = prepare_binary_bank(R)
    if not (0.0 <= tau <= 1.0):
        raise ValueError(f"tau must be in [0, 1]; got {tau}")
    k = validate_finite_k(bank.trial_count, k)
    threshold = max(1, int(math.ceil(tau * k)))
    moments = posterior_moments(
        bank,
        CountScore.threshold_at_k(k, threshold),
        alpha0=alpha0,
        beta0=beta0,
    )
    return moments.mean, float(math.sqrt(moments.variance))


def _mg_pass_at_k_bayes(
    R: np.ndarray, k: int, alpha0: float = 1.0, beta0: float = 1.0
) -> tuple[float, float]:
    """Posterior mean/std for the i.i.d. mG-Pass@k quantity."""
    bank = prepare_binary_bank(R)
    k = validate_finite_k(bank.trial_count, k)
    moments = posterior_moments(
        bank,
        CountScore.mg_at_k(k),
        alpha0=alpha0,
        beta0=beta0,
    )
    return moments.mean, float(math.sqrt(moments.variance))


def g_pass_at_k_ci(
    R: np.ndarray,
    k: int,
    confidence: float = 0.95,
    bounds: tuple[float, float] = (0.0, 1.0),
    alpha0: float = 1.0,
    beta0: float = 1.0,
) -> tuple[float, float, float, float]:
    r"""
    Bayesian posterior summary for G-Pass@k.

    G-Pass@k is the all-success threshold, so this is the same posterior
    target as :func:`~scorio.eval.pass_hat_k_ci`.

    Args:
        R: :math:`M \times N` binary matrix with entries in :math:`\{0,1\}`.
        k: Number of selected samples with ``1 <= k <= N``.
        confidence: credibility level of the interval.
        bounds: ``(lo, hi)`` clipping bounds for the interval.
        alpha0: Beta prior parameter :math:`\alpha_0`.
        beta0: Beta prior parameter :math:`\beta_0`.

    Returns:
        tuple[float, float, float, float]:
            :math:`(\mu,\; \sigma,\; \text{lo},\; \text{hi})`.
    """
    return pass_hat_k_ci(
        R, k, confidence=confidence, bounds=bounds, alpha0=alpha0, beta0=beta0
    )


def g_pass_at_k_tau_ci(
    R: np.ndarray,
    k: int,
    tau: float,
    confidence: float = 0.95,
    bounds: tuple[float, float] = (0.0, 1.0),
    alpha0: float = 1.0,
    beta0: float = 1.0,
) -> tuple[float, float, float, float]:
    r"""
    Bayesian posterior summary for thresholded G-Pass@k.

    The latent target is the probability that at least
    :math:`\lceil \tau k \rceil` of ``k`` i.i.d. samples are correct.

    Args:
        R: :math:`M \times N` binary matrix with entries in :math:`\{0,1\}`.
        k: Number of selected samples with ``1 <= k <= N``.
        tau: Threshold parameter in ``[0, 1]``.
        confidence: credibility level of the interval.
        bounds: ``(lo, hi)`` clipping bounds for the interval.
        alpha0: Beta prior parameter :math:`\alpha_0`.
        beta0: Beta prior parameter :math:`\beta_0`.

    Returns:
        tuple[float, float, float, float]:
            :math:`(\mu,\; \sigma,\; \text{lo},\; \text{hi})`.
    """
    mu, sigma = _g_pass_at_k_tau_bayes(R, k, tau, alpha0=alpha0, beta0=beta0)
    lo, hi = normal_credible_interval(
        mu, sigma, credibility=confidence, two_sided=True, bounds=bounds
    )
    return float(mu), float(sigma), float(lo), float(hi)


def mg_pass_at_k_ci(
    R: np.ndarray,
    k: int,
    confidence: float = 0.95,
    bounds: tuple[float, float] = (0.0, 1.0),
    alpha0: float = 1.0,
    beta0: float = 1.0,
) -> tuple[float, float, float, float]:
    r"""
    Bayesian posterior summary for mG-Pass@k.

    The latent target uses the published discrete threshold weighting from
    ``0.5`` to ``1.0`` implemented by :func:`~scorio.eval.mg_pass_at_k`.
    This weighting approximates, but for odd ``k`` does not exactly equal,
    the corresponding continuous area.

    Args:
        R: :math:`M \times N` binary matrix with entries in :math:`\{0,1\}`.
        k: Number of selected samples with ``1 <= k <= N``.
        confidence: credibility level of the interval.
        bounds: ``(lo, hi)`` clipping bounds for the interval.
        alpha0: Beta prior parameter :math:`\alpha_0`.
        beta0: Beta prior parameter :math:`\beta_0`.

    Returns:
        tuple[float, float, float, float]:
            :math:`(\mu,\; \sigma,\; \text{lo},\; \text{hi})`.
    """
    mu, sigma = _mg_pass_at_k_bayes(R, k, alpha0=alpha0, beta0=beta0)
    lo, hi = normal_credible_interval(
        mu, sigma, credibility=confidence, two_sided=True, bounds=bounds
    )
    return float(mu), float(sigma), float(lo), float(hi)


__all__ = [
    "g_pass_at_k",
    "g_pass_at_k_tau",
    "mg_pass_at_k",
    "g_pass_at_k_ci",
    "g_pass_at_k_tau_ci",
    "mg_pass_at_k_ci",
]

File: js/scorio/src/eval/gpass.ts
/**
 * Generalized pass-family metrics for binary outcomes — G-Pass@k, the
 * thresholded G-Pass@k_τ (at least `ceil(τk)` successes), and mG-Pass@k
 * (the mean over thresholds `τ ∈ [0.5, 1.0]`), with Beta-posterior credible
 * intervals. Port of `scorio/eval/gpass.py`.
 *
 * References: Liu et al. (2024), arXiv:2412.13147; Yao et al. (2024),
 * arXiv:2406.12045.
 */

import {
  betaRatio,
  comb,
  hypergeomPmf,
  hypergeomSf,
  logBetaRatio,
} from "./internal/math.js";
import { normalCredibleInterval, type Bounds } from "./internal/ci.js";
import {
  asMatrix,
  rowSums,
  validateBinary,
  type Matrix,
} from "./internal/validate.js";
import { passAtK, passHatK, passHatKCi } from "./passAtK.js";

function checkK(k: number, N: number): void {
  if (!(k >= 1 && k <= N)) {
    throw new Error(`k must satisfy 1 <= k <= N (N=${N}); got k=${k}`);
  }
}

function checkTau(tau: number): void {
  if (!(tau >= 0.0 && tau <= 1.0)) {
    throw new Error(`tau must be in [0, 1]; got ${tau}`);
  }
}

/** Per-row Beta posterior parameters `[alpha, beta]` for binary outcomes. */
function binaryBetaPosterior(
  Rm: readonly (readonly number[])[],
  alpha0: number,
  beta0: number,
): { alpha: number[]; beta: number[]; N: number } {
  validateBinary(Rm);
  const N = Rm[0]!.length;
  const c = rowSums(Rm);
  return {
    alpha: c.map((ci) => alpha0 + ci),
    beta: c.map((ci) => beta0 + (N - ci)),
    N,
  };
}

/**
 * G-Pass@k: the all-success (`τ = 1`) threshold, an alias for Pass^k.
 *
 * Included for literature using the G-Pass@k naming convention.
 */
export function gPassAtK(R: Matrix, k: number): number {
  return passHatK(R, k);
}

/**
 * G-Pass@k_τ: average probability of at least `ceil(τk)` successes among `k`
 * selected samples (unbiased hypergeometric estimator). `τ = 0` reduces to
 * Pass@k and `τ = 1` to Pass^k.
 */
export function gPassAtKTau(R: Matrix, k: number, tau: number): number {
  const Rm = asMatrix(R);
  validateBinary(Rm);
  const N = Rm[0]!.length;
  checkTau(tau);
  checkK(k, N);
  if (!Number.isInteger(k)) return NaN;

  if (tau <= 0.0) {
    return passAtK(Rm, k);
  }

  const nu = rowSums(Rm);
  const j0 = Math.max(1, Math.ceil(tau * k));
  const M = Rm.length;
  const vals = nu.map((v) => hypergeomSf(N, v, k, j0));
  return vals.reduce((s, v) => s + v, 0) / M;
}

/**
 * mG-Pass@k: the mean generalized pass metric, `2 ∫_{0.5}^{1} G-Pass@k_τ dτ`.
 *
 * The integral over `τ ∈ [0.5, 1.0]` collapses to the closed form
 * `(2/k) · Σ_{j=m+1}^{k} (j - m) · P(X = j)`, where `m = ceil(k/2)` is the
 * majority threshold and `X ~ Hypergeometric(N, ν, k)`.
 */
export function mgPassAtK(R: Matrix, k: number): number {
  const Rm = asMatrix(R);
  validateBinary(Rm);
  const N = Rm[0]!.length;
  checkK(k, N);
  if (!Number.isInteger(k)) {
    throw new TypeError("'number' object cannot be interpreted as an integer");
  }

  const nu = rowSums(Rm);

  const majority = Math.ceil(0.5 * k);
  if (majority >= k) {
    return 0.0;
  }

  const M = Rm.length;
  const vals = new Array<number>(M).fill(0);
  for (let j = majority + 1; j <= k; j++) {
    for (let i = 0; i < M; i++) {
      const v = nu[i]!;
      const pmf = hypergeomPmf(N, v, k, j);
      vals[i]! += (j - majority) * pmf;
    }
  }
  return vals.reduce((s, v) => s + (v * 2.0) / k, 0) / M;
}

/**
 * Posterior mean/std of `Σ_j coeff_j · 1{X >= ...}` for a hypergeometric-style
 * weighted sum whose i.i.d. Beta-Binomial moments are
 * `E[(success-budget) terms] = Beta(a+j, b+k-j)/Beta(a, b)`.
 *
 * The first and second moments need `Beta(a+x, b+k-x)/Beta(a,b)` (`tK`) and
 * `Beta(a+x, b+2k-x)/Beta(a,b)` (`t2K`). Evaluate each required ratio directly
 * in log space. A forward recurrence from x=0 is tempting, but its seed can
 * underflow for large k and then incorrectly leaves every later moment at zero.
 */
function iidWeightedMoments(
  alpha: readonly number[],
  beta: readonly number[],
  M: number,
  js: readonly number[],
  coeff: readonly number[],
  k: number,
): [number, number] {
  let meanSum = 0;
  let varSum = 0;
  for (let i = 0; i < M; i++) {
    const a = alpha[i]!;
    const b = beta[i]!;

    const logTK = new Array<number>(k + 1);
    for (let x = 0; x <= k; x++) logTK[x] = logBetaRatio(a, b, x, k - x);

    const logT2K = new Array<number>(2 * k + 1);
    for (let x = 0; x <= 2 * k; x++) {
      logT2K[x] = logBetaRatio(a, b, x, 2 * k - x);
    }

    let m = 0;
    for (let idx = 0; idx < js.length; idx++) {
      const c = coeff[idx]!;
      if (c !== 0.0) m += Math.exp(Math.log(c) + logTK[js[idx]!]!);
    }

    let e2 = 0;
    for (let idxJ = 0; idxJ < js.length; idxJ++) {
      const cJ = coeff[idxJ]!;
      if (cJ === 0.0) continue;
      const j = js[idxJ]!;
      for (let idxL = 0; idxL < js.length; idxL++) {
        const cL = coeff[idxL]!;
        if (cL === 0.0) continue;
        // Preserve Python's floating-point evaluation order. For sufficiently
        // large k the coefficient product overflows while the beta moment
        // underflows, producing NaN; Python's max(0.0, NaN) then clips that
        // row variance to zero.
        e2 += cJ * cL * Math.exp(logT2K[j + js[idxL]!]!);
      }
    }

    meanSum += m;
    const rawVariance = e2 - m * m;
    varSum += Number.isNaN(rawVariance) ? 0.0 : Math.max(0.0, rawVariance);
  }
  return [meanSum / M, Math.sqrt(varSum) / M];
}

/** Posterior mean/std for the i.i.d. G-Pass@k_τ quantity. */
function gPassAtKTauBayes(
  R: Matrix,
  k: number,
  tau: number,
  alpha0: number,
  beta0: number,
): [number, number] {
  const Rm = asMatrix(R);
  const { alpha, beta, N } = binaryBetaPosterior(Rm, alpha0, beta0);
  checkTau(tau);
  checkK(k, N);

  if (tau <= 0.0) {
    return passAtKBayes(Rm, k, alpha0, beta0);
  }
  if (tau >= 1.0) {
    return passHatKBayes(Rm, k, alpha0, beta0);
  }
  if (!Number.isInteger(k)) {
    throw new TypeError("'number' object cannot be interpreted as an integer");
  }

  const M = Rm.length;
  const j0 = Math.ceil(tau * k);
  const js: number[] = [];
  for (let j = j0; j <= k; j++) js.push(j);
  const coeff = js.map((j) => comb(k, j));

  return iidWeightedMoments(alpha, beta, M, js, coeff, k);
}

/** Posterior mean/std for the i.i.d. mG-Pass@k quantity. */
function mgPassAtKBayes(
  R: Matrix,
  k: number,
  alpha0: number,
  beta0: number,
): [number, number] {
  const Rm = asMatrix(R);
  const { alpha, beta, N } = binaryBetaPosterior(Rm, alpha0, beta0);
  checkK(k, N);
  if (!Number.isInteger(k)) {
    throw new TypeError("'number' object cannot be interpreted as an integer");
  }

  const majority = Math.ceil(0.5 * k);
  if (majority >= k) {
    return [0.0, 0.0];
  }

  const M = Rm.length;
  const js: number[] = [];
  for (let j = majority + 1; j <= k; j++) js.push(j);
  const coeff = js.map((j) => (2.0 / k) * (j - majority) * comb(k, j));

  return iidWeightedMoments(alpha, beta, M, js, coeff, k);
}

/** Posterior mean/std for the i.i.d. Pass@k quantity `1 - (1-p)^k`. */
function passAtKBayes(
  R: Matrix,
  k: number,
  alpha0: number,
  beta0: number,
): [number, number] {
  const Rm = asMatrix(R);
  const { alpha, beta, N } = binaryBetaPosterior(Rm, alpha0, beta0);
  checkK(k, N);
  const M = Rm.length;
  let meanSum = 0;
  let varSum = 0;
  for (let i = 0; i < M; i++) {
    const a = alpha[i]!;
    const b = beta[i]!;
    const eQk = betaRatio(a, b, 0, k);
    const eQ2k = betaRatio(a, b, 0, 2 * k);
    const m = 1 - eQk;
    const e2 = 1 - 2 * eQk + eQ2k;
    meanSum += m;
    varSum += Math.max(0, e2 - m * m);
  }
  return [meanSum / M, Math.sqrt(varSum) / M];
}

/** Posterior mean/std for the i.i.d. Pass^k quantity `p^k`. */
function passHatKBayes(
  R: Matrix,
  k: number,
  alpha0: number,
  beta0: number,
): [number, number] {
  const Rm = asMatrix(R);
  const { alpha, beta, N } = binaryBetaPosterior(Rm, alpha0, beta0);
  checkK(k, N);
  const M = Rm.length;
  let meanSum = 0;
  let varSum = 0;
  for (let i = 0; i < M; i++) {
    const a = alpha[i]!;
    const b = beta[i]!;
    const ePk = betaRatio(a, b, k, 0);
    const eP2k = betaRatio(a, b, 2 * k, 0);
    meanSum += ePk;
    varSum += Math.max(0, eP2k - ePk * ePk);
  }
  return [meanSum / M, Math.sqrt(varSum) / M];
}

/**
 * Bayesian `[mu, sigma, lo, hi]` for G-Pass@k (alias for Pass^k posterior).
 */
export function gPassAtKCi(
  R: Matrix,
  k: number,
  confidence = 0.95,
  bounds: Bounds | null = [0.0, 1.0],
  alpha0 = 1.0,
  beta0 = 1.0,
): [number, number, number, number] {
  return passHatKCi(R, k, confidence, bounds, alpha0, beta0);
}

/** Bayesian `[mu, sigma, lo, hi]` for thresholded G-Pass@k_τ. */
export function gPassAtKTauCi(
  R: Matrix,
  k: number,
  tau: number,
  confidence = 0.95,
  bounds: Bounds | null = [0.0, 1.0],
  alpha0 = 1.0,
  beta0 = 1.0,
): [number, number, number, number] {
  const [mu, sigma] = gPassAtKTauBayes(R, k, tau, alpha0, beta0);
  const [lo, hi] = normalCredibleInterval(mu, sigma, confidence, true, bounds);
  return [mu, sigma, lo, hi];
}

/** Bayesian `[mu, sigma, lo, hi]` for mG-Pass@k. */
export function mgPassAtKCi(
  R: Matrix,
  k: number,
  confidence = 0.95,
  bounds: Bounds | null = [0.0, 1.0],
  alpha0 = 1.0,
  beta0 = 1.0,
): [number, number, number, number] {
  const [mu, sigma] = mgPassAtKBayes(R, k, alpha0, beta0);
  const [lo, hi] = normalCredibleInterval(mu, sigma, confidence, true, bounds);
  return [mu, sigma, lo, hi];
}
